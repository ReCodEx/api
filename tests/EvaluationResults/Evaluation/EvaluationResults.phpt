<?php

include '../../bootstrap.php';

use Tester\Assert;

use App\Helpers\JobConfig\JobConfig;
use App\Helpers\JobConfig\TaskConfig;

use App\Helpers\EvaluationResults\EvaluationResults;
use App\Helpers\EvaluationResults\TaskResult;
use App\Helpers\EvaluationResults\TestResult;

use App\Exceptions\JobConfigLoadingException;
use App\Exceptions\ResultsLoadingException;

class TestEvaluationResults extends Tester\TestCase
{

  static $jobConfig = [
    "submission" => [
      "job-id" => "bla bla bla"
    ],
    "tasks" => [
      [ "task-id" => "W", "type" => TaskConfig::TYPE_INITIATION ],
      [
        "task-id" => "X", "test-id" => "A", "type" => TaskConfig::TYPE_EXECUTION,
        "sandbox" => [ "limits" => [[ "hw-group-id" => "A", "memory" => 123, "time" => 456 ]] ]
      ],
      [ "task-id" => "Y", "test-id" => "A", "type" => TaskConfig::TYPE_EVALUATION ]
    ]
  ];

  public function testMissingParams() {
    $jobConfig = new JobConfig(self::$jobConfig);
    
    Assert::exception(function () use ($jobConfig) {
      new EvaluationResults([], $jobConfig);
    }, ResultsLoadingException::CLASS);
    
    Assert::exception(function () use ($jobConfig) {
      new EvaluationResults([ "job-id" => "ratata" ], $jobConfig);
    }, ResultsLoadingException::CLASS);
    
    Assert::exception(function () use ($jobConfig) {
      new EvaluationResults([
        "job-id" => "bla bla bla"
      ], $jobConfig);
    }, ResultsLoadingException::CLASS);
    
    Assert::exception(function () use ($jobConfig) {
      new EvaluationResults([
        "job-id" => "bla bla bla",
        "results" => NULL
      ], $jobConfig);
    }, ResultsLoadingException::CLASS);
    
    Assert::noError(function () use ($jobConfig) {
      new EvaluationResults([
        "job-id" => "bla bla bla",
        "results" => []
      ], $jobConfig);
    });
    
    Assert::exception(function () use ($jobConfig) {
      new EvaluationResults([
        "job-id" => "bla bla bla",
        "results" => [ [ "a" => "b" ] ]
      ], $jobConfig);
    }, ResultsLoadingException::CLASS);
  }

  public function testInitialisationOK() {
    $jobConfig = new JobConfig(self::$jobConfig);
    $results = new EvaluationResults([
      "job-id" => "bla bla bla",
      "results" => [
        [ "task-id" => "W", "status" => "OK" ],
        [ "task-id" => "X", "status" => "OK" ],
        [ "task-id" => "Y", "status" => "OK" ]
      ]
    ], $jobConfig);

    Assert::true($results->initOK());
  }

  public function testInitialisationFailedBecauseOfSkippedTask() {
    $jobConfig = new JobConfig([
      "submission" => [
        "job-id" => "bla bla bla"
      ],
      "tasks" => [
        [ "task-id" => "A", "type" => TaskConfig::TYPE_INITIATION ],
        [ "task-id" => "B", "type" => TaskConfig::TYPE_INITIATION ]
      ]
    ]);
    $results = new EvaluationResults([
      "job-id" => "bla bla bla",
      "results" => [
        [ "task-id" => "A", "status" => "OK" ],
        [ "task-id" => "B", "status" => "SKIPPED" ]
      ]
    ], $jobConfig);

    Assert::false($results->initOK());
  }

  public function testInitialisationFailedBecauseOfFailedTask() {
    $jobConfig = new JobConfig([
      "submission" => [
        "job-id" => "bla bla bla"
      ],
      "tasks" => [
        [ "task-id" => "A", "type" => TaskConfig::TYPE_INITIATION ],
        [ "task-id" => "B", "type" => TaskConfig::TYPE_INITIATION ]
      ]
    ]);
    $results = new EvaluationResults([
      "job-id" => "bla bla bla",
      "results" => [
        [ "task-id" => "A", "status" => "OK" ],
        [ "task-id" => "B", "status" => "FAILED" ]
      ]
    ], $jobConfig);

    Assert::false($results->initOK());
  }

  public function testInitialisationFailedBecauseOfMissingTaskInitResult() {
    $jobConfig = new JobConfig([
      "submission" => [
        "job-id" => "bla bla bla"
      ],
      "tasks" => [
        [ "task-id" => "A", "type" => TaskConfig::TYPE_INITIATION ],
        [ "task-id" => "B", "type" => TaskConfig::TYPE_INITIATION ]
      ]
    ]);
    $results = new EvaluationResults([
      "job-id" => "bla bla bla",
      "results" => [
        [ "task-id" => "A", "status" => "OK" ]
      ]
    ], $jobConfig);

    Assert::false($results->initOK());
  }


  public function testSimpleGetTestResult() {
    $jobConfig = new JobConfig(self::$jobConfig);
    $initRes = [ "task-id" => "W", "status" => "OK" ];
    $execRes = [
      "task-id" => "X",
      "status" => "OK",
      "sandbox_results" => [
        "exitcode"  => 0,
        "max-rss"   => 19696,
        "memory"    => 100,
        "wall-time" => 0.092,
        "exitsig"   => 0,
        "message"   => "This is a random message",
        "status"    => "OK",
        "time"      => 0.037,
        "killed"    => false
      ]
    ];
    $evalRes = [ "task-id" => "Y", "status" => "OK", "judge_output" => "0.456" ]; 
    $results = new EvaluationResults([ "job-id" => "bla bla bla", "results" => [ $initRes, $evalRes, $execRes ] ], $jobConfig); 
    $testConfig = $jobConfig->getTests()["A"];

    $testResult = $results->getTestResult($testConfig, "A");
    Assert::type(TestResult::CLASS, $testResult);
    Assert::equal("A", $testResult->getId());
    Assert::equal("OK", $testResult->getStatus());
    Assert::equal(TRUE, $testResult->isMemoryOK());
    Assert::equal(TRUE, $testResult->isTimeOK());
    Assert::equal(TRUE, $testResult->didExecutionMeetLimits());
    Assert::equal("0.456", $testResult->getJudgeOutput());
    Assert::equal(0.456, $testResult->getScore());

    Assert::equal(1, count($results->getTestsResults("A")));
  }

}

# Testing methods run
$testCase = new TestEvaluationResults;
$testCase->run();
