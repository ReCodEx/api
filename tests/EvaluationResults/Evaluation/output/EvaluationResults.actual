App\Helpers\EvaluationResults\TestResult::__set_state(/* #49ac */ array(
	'config' => App\Helpers\JobConfig\TestConfig::__set_state(/* #6bd7 */ array(
		'id' => 'A',
		'tasks' => array(
			App\Helpers\JobConfig\TaskConfig::__set_state(/* #68a2 */ array(
				'id' => 'X',
				'type' => 'evaluation',
				'testId' => 'A',
				'data' => array('task-id' => 'X', 'test-id' => 'A', 'type' => 'evaluation'),
			)),
			App\Helpers\JobConfig\TaskConfig::__set_state(/* #c3d8 */ array(
				'id' => 'Y',
				'type' => 'execution',
				'testId' => 'A',
				'data' => array(
					'task-id' => 'Y',
					'test-id' => 'A',
					'type' => 'execution',
					'sandbox' => array(
						'limits' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
					),
				),
			)),
		),
		'executionTask' => App\Helpers\JobConfig\ExecutionTaskConfig::__set_state(/* #c775 */ array(
			'limits' => array(
				'A' => App\Helpers\JobConfig\Limits::__set_state(/* #b76d */ array(
					'data' => array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456),
					'id' => 'A',
					'timeLimit' => 456.0,
					'memoryLimit' => 123,
				)),
			),
			'limitsConfig' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
			'id' => 'Y',
			'type' => 'execution',
			'testId' => 'A',
			'data' => array(
				'task-id' => 'Y',
				'test-id' => 'A',
				'type' => 'execution',
				'sandbox' => array(
					'limits' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
				),
			),
		)),
		'evaluationTask' => /* App\Helpers\JobConfig\TaskConfig dumped on line 5 */,
	)),
	'executionResult' => App\Helpers\EvaluationResults\ExecutionTaskResult::__set_state(/* #cac9 */ array(
		'stats' => App\Helpers\EvaluationResults\Stats::__set_state(/* #bea4 */ array(
			'data' => array(
				'exitcode' => 0,
				'max-rss' => 19696,
				'memory' => 6032,
				'wall-time' => 0.092,
				'exitsig' => 0,
				'message' => 'This is a random message',
				'status' => 'OK',
				'time' => 0.037,
				'killed' => FALSE,
			),
			'time' => 0.037,
			'memory' => 6032,
			'exitcode' => 0,
			'message' => 'This is a random message',
			'killed' => FALSE,
		)),
		'data' => array(
			'task-id' => 'Y',
			'status' => 'OK',
			'sandbox_results' => array(
				'exitcode' => 0,
				'max-rss' => 19696,
				'memory' => 6032,
				'wall-time' => 0.092,
				'exitsig' => 0,
				'message' => 'This is a random message',
				'status' => 'OK',
				'time' => 0.037,
				'killed' => FALSE,
			),
		),
		'id' => 'Y',
		'status' => 'OK',
	)),
	'evaluationResult' => App\Helpers\EvaluationResults\EvaluationTaskResult::__set_state(/* #9e3f */ array(
		'judgeOutput' => '',
		'score' => NULL,
		'data' => array(
			'task-id' => 'Y',
			'status' => 'OK',
			'sandbox_results' => array(
				'exitcode' => 0,
				'max-rss' => 19696,
				'memory' => 6032,
				'wall-time' => 0.092,
				'exitsig' => 0,
				'message' => 'This is a random message',
				'status' => 'OK',
				'time' => 0.037,
				'killed' => FALSE,
			),
		),
		'id' => 'Y',
		'status' => 'OK',
	)),
	'status' => 'FAILED',
	'limits' => /* App\Helpers\JobConfig\Limits dumped on line 27 */,
	'statsInterpretation' => App\Helpers\EvaluationResults\StatsInterpretation::__set_state(/* #3e6c */ array(
		'stats' => /* App\Helpers\EvaluationResults\Stats dumped on line 50 */,
		'limits' => /* App\Helpers\JobConfig\Limits dumped on line 27 */,
	)),
))
