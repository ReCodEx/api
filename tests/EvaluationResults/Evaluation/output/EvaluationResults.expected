App\Helpers\EvaluationResults\TestResult::__set_state(/* #0b4d */ array(
	'config' => App\Helpers\JobConfig\TestConfig::__set_state(/* #6bd7 */ array(
		'id' => 'A',
		'tasks' => array(
			App\Helpers\JobConfig\TaskConfig::__set_state(/* #68a2 */ array(
				'id' => 'X',
				'type' => 'evaluation',
				'testId' => 'A',
				'data' => array('task-id' => 'X', 'test-id' => 'A', 'type' => 'evaluation'),
			)),
			App\Helpers\JobConfig\TaskConfig::__set_state(/* #c3d8 */ array(
				'id' => 'Y',
				'type' => 'execution',
				'testId' => 'A',
				'data' => array(
					'task-id' => 'Y',
					'test-id' => 'A',
					'type' => 'execution',
					'sandbox' => array(
						'limits' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
					),
				),
			)),
		),
		'executionTask' => App\Helpers\JobConfig\ExecutionTaskConfig::__set_state(/* #c775 */ array(
			'limits' => array(
				'A' => App\Helpers\JobConfig\Limits::__set_state(/* #b76d */ array(
					'data' => array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456),
					'id' => 'A',
					'timeLimit' => 456.0,
					'memoryLimit' => 123,
				)),
			),
			'limitsConfig' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
			'id' => 'Y',
			'type' => 'execution',
			'testId' => 'A',
			'data' => array(
				'task-id' => 'Y',
				'test-id' => 'A',
				'type' => 'execution',
				'sandbox' => array(
					'limits' => array(array('hw-group-id' => 'A', 'memory' => 123, 'time' => 456)),
				),
			),
		)),
		'evaluationTask' => /* App\Helpers\JobConfig\TaskConfig dumped on line 5 */,
	)),
	'executionResult' => App\Helpers\EvaluationResults\ExecutionTaskResult::__set_state(/* #4cc1 */ array(
		'stats' => App\Helpers\EvaluationResults\Stats::__set_state(/* #4aad */ array(
			'data' => array(
				'exitcode' => 0,
				'max-rss' => 19696,
				'memory' => 6032,
				'wall-time' => 0.092,
				'exitsig' => 0,
				'message' => 'This is a random message',
				'status' => 'OK',
				'time' => 0.037,
				'killed' => FALSE,
			),
			'time' => 0.037,
			'memory' => 6032,
			'exitcode' => 0,
			'message' => 'This is a random message',
			'killed' => FALSE,
		)),
		'data' => array(
			'task-id' => 'Y',
			'status' => 'OK',
			'sandbox_results' => array(
				'exitcode' => 0,
				'max-rss' => 19696,
				'memory' => 6032,
				'wall-time' => 0.092,
				'exitsig' => 0,
				'message' => 'This is a random message',
				'status' => 'OK',
				'time' => 0.037,
				'killed' => FALSE,
			),
		),
		'id' => 'Y',
		'status' => 'OK',
	)),
	'evaluationResult' => App\Helpers\EvaluationResults\EvaluationTaskResult::__set_state(/* #29f7 */ array(
		'judgeOutput' => '0.456',
		'score' => 0.456,
		'data' => array('task-id' => 'X', 'status' => 'OK', 'judge_output' => '0.456'),
		'id' => 'X',
		'status' => 'OK',
	)),
	'status' => 'FAILED',
	'limits' => /* App\Helpers\JobConfig\Limits dumped on line 27 */,
	'statsInterpretation' => App\Helpers\EvaluationResults\StatsInterpretation::__set_state(/* #8ef9 */ array(
		'stats' => /* App\Helpers\EvaluationResults\Stats dumped on line 50 */,
		'limits' => /* App\Helpers\JobConfig\Limits dumped on line 27 */,
	)),
))
